pkg/frontend/openshiftcluster_delete_test.go

```diff
@@ -0,0 +1,223 @@
+package frontend
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/database"
+	"github.com/Azure/ARO-RP/pkg/database/cosmosdb"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/metrics/noop"
+	mock_database "github.com/Azure/ARO-RP/pkg/util/mocks/database"
+	utiltls "github.com/Azure/ARO-RP/pkg/util/tls"
+	"github.com/Azure/ARO-RP/test/util/listener"
+	"github.com/Azure/ARO-RP/test/util/matcher"
+)
+
+func TestDeleteOpenShiftCluster(t *testing.T) {
+	ctx := context.Background()
+
+	clientkey, clientcerts, err := utiltls.GenerateKeyAndCertificate("client", nil, nil, false, true)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	serverkey, servercerts, err := utiltls.GenerateKeyAndCertificate("server", nil, nil, false, false)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	pool := x509.NewCertPool()
+	pool.AddCert(servercerts[0])
+
+	cli := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: pool,
+				Certificates: []tls.Certificate{
+					{
+						Certificate: [][]byte{clientcerts[0].Raw},
+						PrivateKey:  clientkey,
+					},
+				},
+			},
+		},
+	}
+
+	mockSubID := "00000000-0000-0000-0000-000000000000"
+
+	type test struct {
+		name           string
+		resourceID     string
+		mocks          func(*test, *mock_database.MockAsyncOperations, *mock_database.MockOpenShiftClusters, *mock_database.MockSubscriptions)
+		wantStatusCode int
+		wantAsync      bool
+		wantError      string
+	}
+
+	for _, tt := range []*test{
+		{
+			name:       "cluster exists in db",
+			resourceID: fmt.Sprintf("/subscriptions/%s/resourcegroups/resourceGroup/providers/Microsoft.RedHatOpenShift/openShiftClusters/resourceName", mockSubID),
+			mocks: func(tt *test, asyncOperations *mock_database.MockAsyncOperations, openShiftClusters *mock_database.MockOpenShiftClusters, subscriptions *mock_database.MockSubscriptions) {
+				subscriptions.EXPECT().
+					Get(gomock.Any(), mockSubID).
+					Return(&api.SubscriptionDocument{
+						Subscription: &api.Subscription{
+							State: api.SubscriptionStateRegistered,
+							Properties: &api.SubscriptionProperties{
+								TenantID: "11111111-1111-1111-1111-111111111111",
+							},
+						},
+					}, nil)
+
+				expectAsyncOperationDocumentCreate(asyncOperations, strings.ToLower(tt.resourceID), api.ProvisioningStateDeleting)
+
+				openShiftClusters.EXPECT().
+					Patch(gomock.Any(), strings.ToLower(tt.resourceID), gomock.Any()).
+					DoAndReturn(func(ctx context.Context, key string, f func(doc *api.OpenShiftClusterDocument) error) (*api.OpenShiftClusterDocument, error) {
```

Comment:
This mock was a bit tricky: we need to mock `Patch` and at the same time we still want to call the callback (`f` here) and check the modifications the callback is doing (like updating the document status) because these modifications will get saved into the db.

Looks ugly. I'm open for suggestions ;)

Reply:
This suggests to me that gomock isn't adequate for the database layer and that we might be better of writing a fake database object that (at least partially) conforms to the database interface.  I think this implementation is fine for this PR.

Reply:
opened issue #59 to track